rust/connlib/tunnel/src/dns.rs

```diff
@@ -496,66 +541,82 @@ mod tests {
-    }
+    #[test_case("app.*.example.com", "app.foo.bar.example.com"; "single star does not match two level")]
+    #[test_case("app.*com", "app.foo.com"; "single star does not match dot")]
+    #[test_case("app?com", "app.com"; "question mark does not match dot")]
+    fn domain_pattern_does_not_match(pattern: &str, domain: &str) {
+        let pattern = Pattern::new(pattern).unwrap();
+        let candidate = Candidate::from_str(domain).unwrap();
 
-    #[test]
-    fn wildcard_subdomain_match() {
-        assert!(is_subdomain(&domain("foo.com"), "*.foo.com"));
-        assert!(is_subdomain(&domain("a.foo.com"), "*.foo.com"));
-        assert!(is_subdomain(&domain("a.foo.com"), "*.a.foo.com"));
-        assert!(is_subdomain(&domain("b.a.foo.com"), "*.a.foo.com"));
-        assert!(is_subdomain(&domain("a.b.foo.com"), "*.foo.com"));
-        assert!(!is_subdomain(&domain("afoo.com"), "*.foo.com"));
-        assert!(!is_subdomain(&domain("b.afoo.com"), "*.foo.com"));
-        assert!(!is_subdomain(&domain("bar.com"), "*.foo.com"));
-        assert!(!is_subdomain(&domain("foo.com"), "*.a.foo.com"));
+        let matches = pattern.matches(&candidate);
+
+        assert!(!matches);
     }
+}
 
-    #[test]
-    fn question_mark_subdomain_match() {
-        assert!(is_subdomain(&domain("foo.com"), "?.foo.com"));
-        assert!(is_subdomain(&domain("a.foo.com"), "?.foo.com"));
-        assert!(!is_subdomain(&domain("a.b.foo.com"), "?.foo.com"));
-        assert!(!is_subdomain(&domain("bar.com"), "?.foo.com"));
-        assert!(!is_subdomain(&domain("foo.com"), "?.a.foo.com"));
-        assert!(!is_subdomain(&domain("afoo.com"), "?.foo.com"));
+#[cfg(feature = "divan")]
+mod benches {
+    use super::*;
+    use rand::{distributions::DistString, seq::IteratorRandom, Rng};
+
+    #[divan::bench(
+        consts = [10, 100, 1_000, 10_000, 100_000]
+    )]
+    fn match_domain_linear<const NUM_RES: u128>(bencher: divan::Bencher) {
+        bencher
+            .with_inputs(|| {
+                let mut resolver = StubResolver::new(HashMap::default());
+                let mut rng = rand::thread_rng();
+
+                for n in 0..NUM_RES {
+                    resolver.add_resource(ResourceId::from_u128(n), make_domain(&mut rng));
+                }
+
+                let needle = resolver
+                    .dns_resources
+                    .keys()
+                    .choose(&mut rng)
+                    .unwrap()
+                    .to_string();
+
+                let needle = DomainName::vec_from_str(&needle).unwrap();
+
+                (resolver, needle)
+            })
+            .bench_refs(|(resolver, needle)| resolver.match_resource_linear(needle).unwrap());
     }
 
-    fn domain(name: &str) -> DomainName {
-        DomainName::vec_from_str(name).unwrap()
+    fn make_domain(rng: &mut impl Rng) -> String {
+        (0..rng.gen_range(2..5))
+            .map(|_| rand::distributions::Alphanumeric.sample_string(rng, 3))
```

Comment:
We should bench patterns with glob, I'm worried about someone DoSing a gateway with a huge domain using some address with a wildcard pattern

Reply:
I briefly looked at the implementation and I think it is `O(N)` with the length of the domain. Not sure if that can be done in any more efficient way.